packages/apollo-server-core/src/ApolloServer.ts

```diff
@@ -563,15 +583,49 @@ export class ApolloServerBase {
         )
       ).filter(
         (maybeServerListener): maybeServerListener is GraphQLServerListener =>
-          typeof maybeServerListener === 'object' &&
-          !!maybeServerListener.serverWillStop,
+          typeof maybeServerListener === 'object',
       );
-      this.toDispose.add(async () => {
-        await Promise.all(
-          serverListeners.map(({ serverWillStop }) => serverWillStop?.()),
-        );
+
+      serverListeners.forEach(({ schemaDidChange, serverWillStop }) => {
+        let unsubscribe: Unsubscriber | undefined;
+        if (schemaDidChange) {
+          if (maybeGateway) {
+            // Note that latestApiSchema should have been set at least once by now, specifically
+            // during the call to gateway.load().
+            if (!latestApiSchema) {
+              throw new Error(
+                `Gateway did not notify listeners when loading schema.`,
+              );
+            }
+            // Note that it's important for there to be no await between this schemaDidChange()
+            // invocation and the registration of the new schema change callback, otherwise plugins
+            // could miss schema changes.
+            schemaDidChange({
```

Comment:
I continue to not like the idea of calling schemaDidChange redundantly with the serverDidStart with the same schema. This ties in to what we are discussing with @trevor-scheer over at https://github.com/apollographql/federation/pull/738#discussion_r634948971
I'd really prefer "did change" to actually mean "did change" and not "here is the same information we already gave you again".

If we're concerned about the serverWillStart overlapping with update thing, that makes sense: but I think we should make the change I want in 738 and then instead of asserting that `latestApiSchema` must be set, use whether or not it is set to call schemaDidChange only in the case of this overlap.

Here's a thought: what if in addition to `gateway.onSchemaChange`, we also allowed you to pass an onSchemaChange handler to `gateway.load`, with the semantics that this handler would be installed immediately after the initial load and it's gateway's job to make sure no changes are missed? Would that let us simplify things?

Or hmm.

Hmm.

OK, having looked at the whole PR, I do see that there is something simplifying about having a callback that is called every time the schema is updated, including the first time. Maybe my objection is just to the "changed" terminology. Can we come up with a more precise term that doesn't imply a change from a previous value? `schemaDidLoad` probably implies only the first time.

If we're doing that, we could consider changing #738 so that instead of adding a second argument to the callback, we just add a new `onSchemaWhatever` whose callback takes an object with keys instead of separate arguments.

And ooh! I think this will let us make the ApolloServer schemaDidWhatever take coreSupergraphSdl as a non-`?` argument. The idea is that we can move the runtime check into the schema loading code from the schema reporter. We can basically do:

- If any plugin registered schemaDidWhatever, and there is a gateway, then `gateway.onSchemaWhatever` must exist or it's an immediate startup error.  (ie, it's an error to combine schema reporting with an old gateway, which shouldn't break any existing users since that's already an error.)
- However, if no plugin registered schemaDidWhatever, then we can fall back to `gateway.onSchemaChange` if onSchemaWhatever doesn't exist.

Reply:
I've changed the name to `schemaDidLoadOrUpdate` and changed apollographql/federation#738 to have a new `onSchemaLoadOrUpdate` method that takes an object with keys/properties.

>And ooh! I think this will let us make the ApolloServer schemaDidWhatever take coreSupergraphSdl as a non-`?` argument.

I think that argument still ends up needing to be `?` to be for the non-federated schema reporting case.

>The idea is that we can move the runtime check into the schema loading code from the schema reporter. We can basically do:
>- If any plugin registered schemaDidWhatever, and there is a gateway, then `gateway.onSchemaWhatever` must exist or it's an immediate startup error. (ie, it's an error to combine schema reporting with an old gateway, which shouldn't break any existing users since that's already an error.)

Sounds good, I've added a check for this near the end of `_start()`. Note that I'm beginning to think no error is currently thrown when trying to use the schema reporting plugin with a gateway (or at least, I don't see any code that throws when that's the case). However, our autoreg pipeline silently throws away reports from gateways.

>- However, if no plugin registered schemaDidWhatever, then we can fall back to `gateway.onSchemaChange` if onSchemaWhatever doesn't exist.

We have to register gateway listeners early in `_start()` before we ask plugins whether they provide `schemaDidLoadOrUpdate`, so the new code just uses `schemaDidLoadOrUpdate` if available or falls back to `onSchemaChange` (regardless of plugin registration).